libsolidity/lsp/ReferenceCollector.cpp

```diff
@@ -0,0 +1,192 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+#include <libsolidity/lsp/ReferenceCollector.h>
+#include <libsolidity/lsp/Utils.h>
+
+#include <libsolidity/ast/AST.h>
+#include <libsolidity/lsp/LanguageServer.h>
+
+#include <libsolutil/Common.h>
+
+using namespace solidity::frontend;
+using namespace solidity::langutil;
+using namespace std::string_literals;
+using namespace std;
+
+namespace solidity::lsp
+{
+
+namespace
+{
+
+vector<Declaration const*> allAnnotatedDeclarations(Identifier const* _identifier)
+{
+	vector<Declaration const*> output;
+	output.push_back(_identifier->annotation().referencedDeclaration);
+	output += _identifier->annotation().candidateDeclarations;
```

Comment:
I barely remember a discussion with @chriseth on that matter where I was simply adding referencedDeclarations. But if I remember correctly, he suggested me to include candidateDeclarations as well. I'm fine in removing that part again. But to me it did sounds like useful to be included when looking for referenced of overloads as well, including for symbolic renames, where you want to rename all overloads most likely (if not you can always opt-out client-side).

Reply:
Opt out? How does that work without the server handling it explicitly? Can the client tell which ones are just overloads?

If you can switch between the two it's good, but if you can only have one, finding only the exact thing you're looking for sounds a bit more useful. How do LSPs for other languages with overloading do it?

Reply:
I asked @ekpyron how he would prefer it he also thinks that only seeing the matching overload is more useful.

The typical scenario would be when you're reading the code of a function and wondering what calls this function. In this case seeing the other overloads seems less useful to us. Maybe it's still useful but rather as an option, not necessarily as the default thing to do.

Reply:
I've removed the inclusion of candidates. For reference: How the opt-out works is that solc would respond with a set of references, and the user usually can then (in case of symbolic rename for example) say either rename all or untick the checkbox of symbols the user would like not to rename, or cancel completely. Nevertheless, I've removed the candidates.

Reply:
Having to manually weed out incorrect references sounds really tedious and error-prone to me. That's the dirty work I'd expect the LSP to do for me. Do other LSPs really do that?